tmk_core/common/bootmagic.c

```diff
@@ -39,6 +39,7 @@ void bootmagic(void) {
     /* eeconfig clear */
     if (bootmagic_scan_keycode(BOOTMAGIC_KEY_EEPROM_CLEAR)) {
         eeconfig_init();
+        matrix_init();
```

Comment:
Calling `matrix_init` here and `keyboard_init` when handling `EEPROM_RESET` feels inconsistent. Architecturally, invoking a core part of the init path outside the init path without any form of deinit sounds like the wrong thing to do.

For instance, what happens with assumptions that `matrix_init_kb` or `matrix_init_user` is only called on startup? (maybe configuring a piece of hardware that cannot be reconfigured later, allocating memory. Not 100% on an existing use case at the moment)

Reply:
I definitely agree with you on the inconsistency. 

However, calling `keyboard_init` here would lead to a situation where the keyboard could actually get stuck in an infinite loop, since bootmagic full is called from keyboard_init

Also, to muddle things, where things are init-ed is very inconsistent, which makes it hard to use just a single function for this, without changing a lot.   

Admittedly, this is putting off tech debt, but short of a massive change, (which I think needs to happen anyways), I'm not sure of a good way to implement that here.


but leaving the keyboard_init in the eeconfig_init isn't a good solution, either.  It has some ... weird behavior.  I've seen my boards (which were using this code) re-init the keyboard 2, 3, and sometimes 4 times, because of it. This is a lighter touch, as it shouldn't trigger it everywhere.

Reply:
>Admittedly, this is putting off tech debt, but short of a massive change, 

Its worth noting that this is still adding to that technical debt, now the init order is muddy.

matrix_init being called twice, calls debounce_init again, which can do https://github.com/qmk/qmk_firmware/blob/7372ce6afdb1030463e09cef92e0d7befbd1d474/quantum/debounce/eager_pr.c#L44 without freeing the memory. Thats quite bad....